README.md

```diff
@@ -0,0 +1,133 @@
+In our example: if the `find` action fails (so no coke is found), the `bring` action will not be executed.
+
+## The Action life cycle
+
+The Action life cycle consists of the following phases:
+ - Instantiation:
+    - initializing member variables (explained further in `example_action`):
+      - `_config_result`
+      - `_execute_result`
+      - `_required_field_prompts`
+      - `_required_passed_knowledge`
+      - `_required_skills`
+      - `_knowledge`
+    - the Action's implementation may add checks for static resources. E.g.
+      - knowledge availability
+      - server availability
+      - etc.
+ - Configuration
+    - checking task semantics using `_required_field_prompts`
+    - checking passed knowledge using using `_required_passed_knowledge`
+    - checking robot for required skills using `_required_skills`
+    - calling the implementation's `_configure`
+ - Execution
+    - performing the actual action
+
+## Implementing your own Action
+Basic steps for implementing an action are as follows:
+  - Create a class that inherits from the Action base class [here](action_server/src/action_server/actions):
+    - implement the \_configure method
+    - implement the \_start method
+    - implement the \_cancel method
+  - Add your new action to the actions module [here](action_server/src/action_server/actions/__init__.py).
+  - Add a command resolving to the required semantics to the grammar you want to test your action with, and test your Action using the natural language console configured with that grammar.
+
+For a detailed guide on the internals of an Action implementation and instructions on how to implement your own Action, take a look at the example implementation: `example_action`.
+This contains a lot of documentation and explanation on the details of Actions.
+
+## FAQ
+
+ - **Why json?**
+   - I know, it hurts to put a [json](https://en.wikipedia.org/wiki/JSON) string in a ROS message.
+   But every action has its own semantics, its own parameters and its own structure in these parameters.
+   This means it doesn't fit in a static ROS message a client can send to the action server. The ugliest part is: there is no clear interface definition between client and server.
+   The client side normally uses a grammar to parse natural language commands, so the grammar specifies the client side of the interface.
+   On the server side, the action specifies the required structure of the semantics.
+   So the Action implementation defines the server side of the interface.
+   It would be nice if we could at least verify that these structures agree on startup, instead of getting weird behavior or crashes during task execution.
+   One of the open issues for improvement is to use [json schemas](http://json-schema.org/) to let the Action implementation specify the interface.
+   We can then let the client query this interface specification, so that it can verify its source of task semantics.
```

Comment:
Several lines actually belong in the issue rather than here, and it's fine to refer to that issue for further reading.

I would leave it at something like 'Unfortunately, each action has its own structure, to make the information passable over a ROS network connection, the only simple option is to drop down to nested lists and dicts with strings, ie a json format. For detailed discussion see issue #23 .'